### DISCUSSION TITLE
Removal of "allow_incomplete_nzb" Special setting

### DISCUSSION THREAD

**Discussion by Shadowhwk** [sabnzbd/sabnzbd#2201] [General]
Hello,

I received a notification with the last update of the container I'm using on unRAID that there are plans to remove the "allow_incomplete_nzb" option under special settings.

The reason I have this enabled is for the very rare cases that I try to obtain content from multiple sources when there are no single complete options via nzb or torrent. In some cases that might mean trying .nzb files which are known to be incomplete.

If this function were to be removed, it won't have a major impact on my usage.

**Comment by Safihre** [comment]
I don't think you understand what the option does. This option is only about when the `.nzb` *file* itself is corrupt, so not when the download is incomplete on the server.

**Reply by Shadowhwk** [reply]
I'm doing the same thing as jlhollin mentioned, but also won't be upset if it's removed. I can always use Forté Agent to try and manually grab items that have corrupted .nzb files.

**Comment by jlhollin** [comment]
I set this for the same reason as Shadowhwk. I'm often fetching the very oldest files on a server and sometimes nzb files from indexers or on usenet itself are incomplete. In the case where part of the file is on one host and more of it is on another host I pull the files into notepad++ and hand-edit it to make one as complete as possible. Even an incomplete resuling file may have enough blocks for the par2 files to reconstruct it. But I need the incomplete nzbs to do that.

But it wouldn't be the end of the world if Sabnzbd didn't support it. I would just have to use Pan or XNews or some other lower-level newsreader to try to get the incomplete files.

**Comment by Safihre** [comment]
@jlhollin So you really have files that are corrupt but you still can and want to use?
Interesting.

**Reply by Shadowhwk** [reply]
That's what using the corrupted .nzb allows us to fix. As an example we have corrupted a.nzb which was able to grab all but 2 parts. It's not an incomplete upload, it's actual corruption of the .nzb (which sometimes requires manual editing). I then find corrupted b.nzb which is the same title, but is able to grab the 2 parts needed for a.nzb to successfully extract. Yes, more often it's not a corrupt .nzb, but an incomplete post on the usenet provider. I do have multiple backbones and multiple indexers so it's usually not an issue finding a non-corrupt .nzb. It's pretty rare that this functionality would get used, so as stated before I won't be disappointed if it gets removed.

**Reply by thezoggy** [reply]
when safihre talks about incomplete nzb its malformed xml. not a partial download/incomplete download/etc.  

The nzb this is about handling is when sab goes to parse the xml and its incomplete.. which normally would get rejected and skipped.
Which then throws the sab log "Invalid NZB file " with a traceback.

**Reply by Safihre** [reply]
Indeed, you would see in the interface the `WARNING: Invalid NZB file %s, skipping (error: %s)` (with something filled for `%s`).
Is this indeed what you have seen, ever? Also @jlhollin

**Reply by pl77** [reply]
I appreciate the clarification of this feature from @thezoggy and @Safihre. I also have hand-edited nzbs, particularly those from sources like nzbking, where the regex used to parse the items was inappropriate and I need to reconstruct an nzb by hand or programatically. 

But I haven't (knowingly) downloaded incomplete xmls that wouldn't parse because they were chopped off halfway through, just xmls that were incorrect for the subject I was trying to grab.

**Reply by NobodySpecial25** [reply]
Quite simply there are ways of repairing files by Frankensteining other posts and all other creative means. Let me figure out if the post is unusable, SABnzb isn't infallible. What's wrong with keeping the option of disabling this feature. If it wasn't for SABnzb's ability to deal with obfuscated posts, I'd use something else.

**Reply by albino1** [reply]
If you reconstruct or Frankenstein an NZB together is still "incomplete" or "corrupt" according to SABnzbd? 

I've definitely done similar things in the past for similar reasons, like broken pieces split across NZBKing/Binsearch/NZBIndex.  I can't remember if SAB throws an error, but even if it did I would just set it to `Download` only and not allow it to `+Repair`/`+Unpack`/`+Delete` so that I could then manually deal with the files myself, which I would think you would have to do regardless of this setting.

**Reply by thezoggy** [reply]
as long as the xml is able to be parsed, then sab wouldnt complain. the option is just how to handle those things it cant parse (because of malform xml / corrupt files / whatever the reason). a big example is when sites are down for maintenance or having issues, they sometimes return the site html or error code in the actual nzb requested. so you end up with garbage in the .nzb

**Reply by snottyone** [reply]
enabled here for much the same reason as mentioned above, the (very very) occasional borked nzb, scraping together a post from multiple servers, etc.
so: at some point the setting was actually useful to me, but it won't kill me if you pull it.

**Reply by Safihre** [reply]
@snottyone Again the same question that I ask everyone: is the borked NZB really a **damaged** file? 
_Or_ do you modify the NZB to add/remove manually files that were missing from the file.

That's a *big* difference, because `allow_incomplete_nzb` only makes sense in case the NZB **file** really is damaged, for example looks like this:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE nzb PUBLIC "-//newzBin//DTD NZB 1.1//EN" "http://www.newzbin.com/DTD/nzb/nzb-1.1.dtd">
<nzb xmlns="http://www.newzbin.com/DTD/2003/nzb">

<head>
 <meta type="category">Movies &gt; HD</meta>
 <meta type="name">Test.1080p.WEB-DL.DD 5.1.H.264-KOGi</meta>
</head>

<file poster="9bc6f720d16c7c7a9857108247cc7b50 &lt;0F6WaGq@rm0d.ETHi&gt;" date="1653651208" subject="&quot;0d7b426a5e2153d7d3374ab588776411.10&quot; yEnc (1/73)">
 <groups>
  <group>alt.binaries.mom</group>
 </groups>
 <segments>
  <segment bytes="792633" number="1">6290b707$4$18125$a92e2ab9@news.vipernews.com</segment>
  <segment bytes="792625" number="2">6290b708$0$14965$a92e2ab9@
```
So it suddenly stops.

**Reply by albino1** [reply]
I think maybe there's some terminology confusion going on here because "incomplete" can mean different things, and the setting itself doesn't mean an NZB is "incomplete" in the way that most users would think of it.  Maybe if people posted some example NZBs Safihre could examine them and quickly clear up the confusion.

**Reply by pl77** [reply]
I agree with both @Safihre and @albino1, that it's rare to encounter unparseable nzbs which lack the correct closing tags because they were interrupted during transmission.  That's a very uncommon issue, whereas dealing with nzbs that lack all the appropriate parts for a post and having to frankenstein them with data from similar nzbs *is* an issue, but not relevant at all to this setting. 

Since I rarely have encountered malformed xml inside nzbs, I'm supportive of removing this feature.  Forcing SAB to parse broken XML may end up introducing security vulnerabilities and since it's so rare to encounter such a file I'm in favour of deprecating this feature. 

EDIT: one option might be simply dumping the unparseable XML/NZB into an 'error/broken_nzb' folder so that people who wish to peer inside can do so at their leisure?  That way you don't have to deal with parsing busted XML inside sab itself.  Just a thought.

**Comment by Shadowhwk** [comment]
Just like others have said, it allows us to take corrupted .nzb files to edit and grab what we can, sometimes requiring grabs from other nzb posts. It's still quite rare for me to do though... maybe twice a year?

**Reply by Safihre** [reply]
So there are really sites that corrupt `.nzb` files halfway through?
Yes, I really have a hard time to imagine that :)

**Comment by jlhollin** [comment]
So...

Is allow_incomplete_nzb simply about XML Schema (or DTD) validation? A file truncated along the way from a sender to a server to an indexer to a backup tape through a restore on the indexer to me, does sometimes happen. Especially for very old files.

There are lots of ways a damaged NZB file could fail validation - from simple things like missing tags to complex things like if the DTD specifies that the contents of the  number attribute in a segment must be present, unique and in sequential order within the parent segments.

Many to most of these can be hand-edited given the bad XML (NZB). 

But it sounds like I'm defending this, and as I already said this is a rare occurrence which SAB doesn't really need to cater for. If you could display a meaningful error (maybe just drop in the message text from the XML validator) and leave the badness in the _ADMIN_ folder of the incomplete directory that would satisfy everyone's mild desire.

**Reply by Safihre** [reply]
You are right indeed, it's _only_ about xml validation. And an error is always displayed where we gave up parsing the NZB file.
The difference is that with this option turned on, the job will be allowed to enter the queue, even if we could only partially parse the NZB file (error will still be thrown).
So when you hand edit the NZB file after that, you still have to remove the broken NZB from the queue and add your fixed version.
